hybridq/noise/channel/channel.py

```diff
@@ -184,23 +184,84 @@ def MatrixChannel(LMatrices: tuple[array, ...],
 
-    ## # Define sample
-    ## def __sample__(self, size: int = None, replace: bool = True):
-    ##     from hybridq.utils import isintegral
-    ##
-    ##     # It is assumed here that s.ndim == 1
-    ##     assert (s.ndim == 1)
-    ##
-    ##     # Get number of elements
-    ##     idxs = np.random.choice(range(len(s)), size=size, replace=replace, p=s)
-    ##
-    ##     # Get gates (it is assumed here that left and right gates are the same)
-    ##     return self.Kraus.gates[0][idxs] if isintegral(idxs) else tuple(
-    ##         self.Kraus.gates[0][x] for x in idxs)
-    ##
-    ## # Define apply
-    ## def __apply__(self, psi, order):
-    ##     raise NotImplementedError
+    # Define sample
+    def __sample__(self, size: int = None, replace: bool = True):
+        from hybridq.utils import isintegral
+
+        # It is assumed here that s.ndim == 1
+        assert (self.s.ndim == 1)
+
+        # Get number of elements
+        idxs = np.random.choice(range(len(self.s)),
+                                size=size,
+                                replace=replace,
+                                p=self.s)
+
+        # Get gates (it is assumed here that left and right gates are the same)
+        return self.Kraus.gates[0][idxs] if isintegral(idxs) else tuple(
+            self.Kraus.gates[0][x] for x in idxs)
+
+    # Define apply
+    def __apply__(self, psi, order):
+        from hybridq.utils import dot
+
+        # It is assumed here that s.ndim == 1
+        assert (self.s.ndim == 1)
+
+        # Define how to get projection
+        def _get_projection(idx):
+            # Get s
+            s = self.s[idx]
+
+            # Get Kraus operator
+            gate = self.Kraus.gates[0][idx]
+
+            # Get projection
+            proj = dot(a=gate.matrix(),
+                       b=psi,
+                       axes=tuple(map(order.index, gate.qubits)),
+                       b_as_complex_array=not np.iscomplexobj(psi))
+
+            # Get normalization
+            norm = np.linalg.norm(proj.ravel())
+
+            # Get probability
+            prob = s * norm**2
+
+            # Return projection, norm and probability
+            return proj, norm, prob
+
+        # Convert order to tuple
+        order = tuple(order)
+
+        # Get random number
+        r = np.random.random()
+
+        # Initialize cumulative
+        c = 0
+
+        # For each Kraus operator (escluding the last one) ...
+        for idx in self.__LMatrices_order[:-1]:
+            # Get projection, norm and probability
+            proj, norm, prob = _get_projection(idx)
+
+            # Update cumulative
+            c += prob
+
+            # If the random number if smaller than the cumulative, break
+            if c >= r:
+                break
+
+        # Otherwise, return the last projection
+        else:
+            # Get projection, norm and probability
+            proj, norm, _ = _get_projection(idx)
```

Comment:
This can be problematic.

If we are unlucky, `r` can be very close to 1, and we might not hit our cutoff (due to rounding errors etc in the norm sum). In this case, this code would use the last Kraus operator in the list. However, it can be that the last one actually gives you a zero vector (i.e. norm=0), and the code will break.

As a simple example, imagine your Kraus operators are |0><0| and |1><1|, and your state is |ψ>=|0>. Then clearly we should apply the first Kraus operator, and return the state |0>. However, if e.g. r=0.999999999, and we compute norm |<Ψ|Ψ>|^2=0.999999 (from rounding error, not perfectly 1), then this code would apply the 1 projector on the state, resulting in the zero vector, and everything will break.

This is why (as I mentioned elsewhere) this safest thing to do is to keep track of the Kraus operator with the corresponding largest `prob`. In the rare case that the cutoff is not satisfied, we should sample that one.

Reply:
I can make this change if you prefer.

Reply:
I also think this should be `idx+1` (since `idx` is the last but one operator ?)

Reply:
actually, maybe it should be `_get_projection(self.__LMatrices_order[-1])` ?